### ISSUE TITLE
Gosu - Failed to install Gem into Fedora 32 Silverblue

### ISSUE THREAD

**Issue by Fulgurance** [gosu/gosu#530]
Hello, today, i try to install gosu gem into the last Fedora Silverblue, but i have this error:

```
[zohran@msi-gs73vr-6rf ~]$ gem install gosu
Building native extensions. This could take a while...
ERROR:  Error installing gosu:
	ERROR: Failed to build gem native extension.

    current directory: /var/home/zohran/.gem/ruby/gems/gosu-0.15.1/ext/gosu
/usr/bin/ruby -I /usr/share/rubygems -r ./siteconf20200507-2619-mqafhq.rb extconf.rb
The Gosu gem requires some libraries to be installed system-wide.
See the following site for a list:
https://github.com/gosu/gosu/wiki/Getting-Started-on-Linux
checking for -lopenal... yes
checking for AL/al.h... yes
creating Makefile

current directory: /var/home/zohran/.gem/ruby/gems/gosu-0.15.1/ext/gosu
make "DESTDIR=" clean
rm -f 
rm -f gosu.so  *.o  *.bak mkmf.log .*.time

current directory: /var/home/zohran/.gem/ruby/gems/gosu-0.15.1/ext/gosu
make "DESTDIR="
g++ -I. -I/usr/include -I/usr/include/ruby/backward -I/usr/include -I../../src -I../..  -I/usr/include/SDL2 -I/usr/include/AL   -I/usr/include/freetype2 -I/usr/include/libpng16 -DHAVE_AL_AL_H    -fPIC -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -std=gnu++11  -D_REENTRANT     -m64 -o Audio.o -c ../../src/Audio.cpp
Dans le fichier inclus depuis ../../src/Audio.cpp:2:
../../src/OggFile.hpp:11:32: attention: option inconnue après « #pragma GCC diagnostic »; vouliez-vous utiliser « -Wcomment » ? [-Wpragmas]
   11 | #pragma GCC diagnostic ignored "-Wcomma"
      |                                ^~~~~~~~~
Dans le fichier inclus depuis ../../src/Audio.cpp:16:
../../src/MPEGFile.hpp: Dans la fonction membre « virtual std::size_t Gosu::MPEGFile::read_data(void*, std::size_t) »:
../../src/MPEGFile.hpp:72:33: attention: comparaison d'expressions entières de types signés différents: « off_t » {aka « long int »} et « std::size_t » {aka « long unsigned int »} [-Wsign-compare]
   72 |                    && position_ != contents_.size()) {
      |                       ~~~~~~~~~~^~~~~~~~~~~~~~~~~~~
Dans le fichier inclus depuis /usr/include/c++/10/cassert:44,
                 depuis ../../src/Audio.cpp:8:
../../src/MPEGFile.hpp: Dans la fonction membre « virtual void Gosu::MPEGFile::rewind() »:
../../src/MPEGFile.hpp:87:49: attention: comparaison d'expressions entières de types signés différents: « off_t » {aka « long int »} et « std::size_t » {aka « long unsigned int »} [-Wsign-compare]
   87 |             assert (position_ >= 0 && position_ <= contents_.size());
      |                                       ~~~~~~~~~~^~~~~~~~~~~~~~~~~~~
g++ -I. -I/usr/include -I/usr/include/ruby/backward -I/usr/include -I../../src -I../..  -I/usr/include/SDL2 -I/usr/include/AL   -I/usr/include/freetype2 -I/usr/include/libpng16 -DHAVE_AL_AL_H    -fPIC -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -std=gnu++11  -D_REENTRANT     -m64 -o AudioImpl.o -c ../../src/AudioImpl.cpp
../../src/AudioImpl.cpp: Dans la fonction « Gosu::Channel Gosu::allocate_channel() »:
../../src/AudioImpl.cpp:64:23: attention: comparaison d'expressions entières de types signés différents: « int » et « const unsigned int » [-Wsign-compare]
   64 |     for (int i = 1; i < CHANNELS; ++i) {
      |                     ~~^~~~~~~~~~
../../src/AudioImpl.cpp: Dans la fonction « bool Gosu::channel_expired(int, int) »:
../../src/AudioImpl.cpp:80:35: attention: comparaison d'expressions entières de types signés différents: « int » et « const unsigned int » [-Wsign-compare]
   80 |     return channel < 0 || channel >= CHANNELS || _tokens[channel] != token;
      |                           ~~~~~~~~^~~~~~~~~~~
../../src/AudioImpl.cpp: Dans la fonction « ALuint Gosu::al_source_for_channel(int) »:
../../src/AudioImpl.cpp:85:32: attention: comparaison d'expressions entières de types signés différents: « int » et « const unsigned int » [-Wsign-compare]
   85 |     if (channel < 0 || channel >= CHANNELS) {
      |                        ~~~~~~~~^~~~~~~~~~~
g++ -I. -I/usr/include -I/usr/include/ruby/backward -I/usr/include -I../../src -I../..  -I/usr/include/SDL2 -I/usr/include/AL   -I/usr/include/freetype2 -I/usr/include/libpng16 -DHAVE_AL_AL_H    -fPIC -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -std=gnu++11  -D_REENTRANT     -m64 -o Bitmap.o -c ../../src/Bitmap.cpp
g++ -I. -I/usr/include -I/usr/include/ruby/backward -I/usr/include -I../../src -I../..  -I/usr/include/SDL2 -I/usr/include/AL   -I/usr/include/freetype2 -I/usr/include/libpng16 -DHAVE_AL_AL_H    -fPIC -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -std=gnu++11  -D_REENTRANT     -m64 -o BitmapIO.o -c ../../src/BitmapIO.cpp
../../src/BitmapIO.cpp:15:32: attention: option inconnue après « #pragma GCC diagnostic »; vouliez-vous utiliser « -Wcomment » ? [-Wpragmas]
   15 | #pragma GCC diagnostic ignored "-Wcomma"
      |                                ^~~~~~~~~
../../src/BitmapIO.cpp: Dans la fonction « int {anonymous}::read_callback(void*, char*, int) »:
../../src/BitmapIO.cpp:32:36: attention: comparaison d'expressions entières de types signés différents: « int » et « size_t » {aka « long unsigned int »} [-Wsign-compare]
   32 |         size_t actual_size = (size < remaining ? size : remaining);
      |                               ~~~~~^~~~~~~~~~~
../../src/BitmapIO.cpp: Au niveau global:
../../src/BitmapIO.cpp:96:32: attention: option inconnue après « #pragma GCC diagnostic »; vouliez-vous utiliser « -Wcomment » ? [-Wpragmas]
   96 | #pragma GCC diagnostic ignored "-Wcomma"
      |                                ^~~~~~~~~
g++ -I. -I/usr/include -I/usr/include/ruby/backward -I/usr/include -I../../src -I../..  -I/usr/include/SDL2 -I/usr/include/AL   -I/usr/include/freetype2 -I/usr/include/libpng16 -DHAVE_AL_AL_H    -fPIC -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -std=gnu++11  -D_REENTRANT     -m64 -o BlockAllocator.o -c ../../src/BlockAllocator.cpp
g++ -I. -I/usr/include -I/usr/include/ruby/backward -I/usr/include -I../../src -I../..  -I/usr/include/SDL2 -I/usr/include/AL   -I/usr/include/freetype2 -I/usr/include/libpng16 -DHAVE_AL_AL_H    -fPIC -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -std=gnu++11  -D_REENTRANT     -m64 -o Channel.o -c ../../src/Channel.cpp
g++ -I. -I/usr/include -I/usr/include/ruby/backward -I/usr/include -I../../src -I../..  -I/usr/include/SDL2 -I/usr/include/AL   -I/usr/include/freetype2 -I/usr/include/libpng16 -DHAVE_AL_AL_H    -fPIC -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -std=gnu++11  -D_REENTRANT     -m64 -o Color.o -c ../../src/Color.cpp
g++ -I. -I/usr/include -I/usr/include/ruby/backward -I/usr/include -I../../src -I../..  -I/usr/include/SDL2 -I/usr/include/AL   -I/usr/include/freetype2 -I/usr/include/libpng16 -DHAVE_AL_AL_H    -fPIC -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -std=gnu++11  -D_REENTRANT     -m64 -o DirectoriesApple.o -c ../../src/DirectoriesApple.cpp
g++ -I. -I/usr/include -I/usr/include/ruby/backward -I/usr/include -I../../src -I../..  -I/usr/include/SDL2 -I/usr/include/AL   -I/usr/include/freetype2 -I/usr/include/libpng16 -DHAVE_AL_AL_H    -fPIC -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -std=gnu++11  -D_REENTRANT     -m64 -o DirectoriesUnix.o -c ../../src/DirectoriesUnix.cpp
g++ -I. -I/usr/include -I/usr/include/ruby/backward -I/usr/include -I../../src -I../..  -I/usr/include/SDL2 -I/usr/include/AL   -I/usr/include/freetype2 -I/usr/include/libpng16 -DHAVE_AL_AL_H    -fPIC -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -std=gnu++11  -D_REENTRANT     -m64 -o DirectoriesWin.o -c ../../src/DirectoriesWin.cpp
g++ -I. -I/usr/include -I/usr/include/ruby/backward -I/usr/include -I../../src -I../..  -I/usr/include/SDL2 -I/usr/include/AL   -I/usr/include/freetype2 -I/usr/include/libpng16 -DHAVE_AL_AL_H    -fPIC -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -std=gnu++11  -D_REENTRANT     -m64 -o FileUnix.o -c ../../src/FileUnix.cpp
../../src/FileUnix.cpp: Dans la fonction membre « virtual void Gosu::File::resize(size_t) »:
../../src/FileUnix.cpp:76:14: attention: la valeur retournée par « int ftruncate(int, __off_t) » est ignorée alors qu'elle est déclarée avec l'attribut « warn_unused_result » [-Wunused-result]
   76 |     ftruncate(pimpl->fd, new_size);
      |     ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~
../../src/FileUnix.cpp: Dans la fonction membre « virtual void Gosu::File::read(size_t, size_t, void*) const »:
../../src/FileUnix.cpp:88:15: attention: la valeur retournée par « ssize_t read(int, void*, size_t) » est ignorée alors qu'elle est déclarée avec l'attribut « warn_unused_result » [-Wunused-result]
   88 |         ::read(pimpl->fd, dest_buffer, length);
      |         ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../src/FileUnix.cpp: Dans la fonction membre « virtual void Gosu::File::write(size_t, size_t, const void*) »:
../../src/FileUnix.cpp:96:12: attention: la valeur retournée par « ssize_t write(int, const void*, size_t) » est ignorée alors qu'elle est déclarée avec l'attribut « warn_unused_result » [-Wunused-result]
   96 |     ::write(pimpl->fd, source_buffer, length);
      |     ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
g++ -I. -I/usr/include -I/usr/include/ruby/backward -I/usr/include -I../../src -I../..  -I/usr/include/SDL2 -I/usr/include/AL   -I/usr/include/freetype2 -I/usr/include/libpng16 -DHAVE_AL_AL_H    -fPIC -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -std=gnu++11  -D_REENTRANT     -m64 -o FileWin.o -c ../../src/FileWin.cpp
g++ -I. -I/usr/include -I/usr/include/ruby/backward -I/usr/include -I../../src -I../..  -I/usr/include/SDL2 -I/usr/include/AL   -I/usr/include/freetype2 -I/usr/include/libpng16 -DHAVE_AL_AL_H    -fPIC -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -std=gnu++11  -D_REENTRANT     -m64 -o Font.o -c ../../src/Font.cpp
../../src/Font.cpp: Dans la fonction membre « void Gosu::Font::set_image(std::string, unsigned int, const Gosu::Image&) »:
../../src/Font.cpp:164:15: erreur: « invalid_argument » n'a pas été déclaré dans cette portée
  164 |         throw invalid_argument("Could not compose '" + codepoint + "' into a single codepoint");
      |               ^~~~~~~~~~~~~~~~
make: *** [Makefile:237: Font.o] Error 1

make failed, exit code 2

Gem files will remain installed in /var/home/zohran/.gem/ruby/gems/gosu-0.15.1 for inspection.
Results logged to /var/home/zohran/.gem/ruby/extensions/x86_64-linux/2.7.0/gosu-0.15.1/gem_make.out
```

**Comment by Fulgurance** [comment]
I have tested into standard fedora, the same problem. Something is wrong into that gem

**Comment by pmer** [comment]
The site https://en.cppreference.com/w/cpp/error/invalid_argument says `class invalid_argument` is defined in `<stdexcept>`.

Perhaps this file was implicitly #included in earlier versions of libstdc++ from a different standard library header that Font.cpp happened to include? Then the latest version of libstdc++ changed that?

Several source files include \<stdexcept\> already. Here is the list of remaining files that use invalid_argument but which don't (directly) include \<stdexcept\>:

- Font.cpp
- Graphics.cpp
- Text.cpp

After these includes are added, it can be tested on Silverblue to see if there are any other exception classes besides invalid_argument that break the build.

**Comment by Fulgurance** [comment]
...hum ... i have checked, libstdc++ and libstdc++-devel packages are already installed.

**Comment by Fulgurance** [comment]
It's big problem to the future of Gosu. Sorry, but i have read into GCC 10, stdexcept was definitively removed. Gosu need an update

**Comment by jlnr** [comment]
If Font.cpp uses a standard exception but does not `#include <stdexcept>`, then that should be definitely be added. Good opportunity for a first pull request - hint hint ;) (I won't have time for another couple of days)

**Comment by phylor** [comment]
Great that this is fixed!

@jlnr Any chance for a `0.15.2` release soon? `0.15.1` currently does not install with gcc 10 because of this issue. `master` compiles fine though, as it includes this fix.

Alternatively, is there a way to install Gosu `master` with bundler? `gem 'gosu', git: 'https://github.com/gosu/gosu.git', branch: :master` doesn't work as the repo doesn't include a gemspec.

**Comment by jlnr** [comment]
@phylor Sorry for the long delay, I just released Gosu 0.15.2.

**Comment by phylor** [comment]
@jlnr Works for me now, thank you very much! :+1: :tada:
